mettle/src/process.c

```diff
@@ -0,0 +1,472 @@
+	free(process);
+}
+
+static int switch_user(const char *user)
+{
+	struct passwd pwd, *result;
+	int bufsize = 8192;
+	void *buf = NULL;
+	int rc;
+	do {
+		buf = realloc(buf, bufsize);
+		if (buf == NULL) {
+			return -1;
+		}
+		rc = getpwnam_r(user, &pwd, buf, bufsize, &result);
+		bufsize += 8192;
+	} while (rc == ERANGE && bufsize < 1048576);
+
+	if (rc == 0) {
+		if (setuid(pwd.pw_uid))
+			exit(1);
+
+		if (setgid(pwd.pw_gid))
+			exit(1);
+	}
+
+	free(buf);
+	return rc;
+}
+
+char ** argv_split(char *args, char **argv, size_t *argc)
+{
+	char *p, *start_of_word;
+	int c;
+	enum states { DULL, IN_WORD, IN_STRING } state = DULL;
+
+	for (p = args; *p != '\0'; p++) {
+		c = (unsigned char) *p;
+		switch (state) {
+		case DULL:
+			if (isspace(c)) {
+				continue;
+			}
+
+			if (c == '"') {
+				state = IN_STRING;
+				start_of_word = p + 1;
+				continue;
+			}
+			state = IN_WORD;
+			start_of_word = p;
+			continue;
+
+		case IN_STRING:
+			if (c == '"') {
+				*p = 0;
+				argv = realloc(argv, sizeof(char *) * (*argc + 1));
+				argv[(*argc)++] = start_of_word;
+				state = DULL;
+			}
+			continue;
+
+		case IN_WORD:
+			if (isspace(c)) {
+				*p = 0;
+				argv = realloc(argv, sizeof(char *) * (*argc + 1));
+				argv[(*argc)++] = start_of_word;
+				state = DULL;
+			}
+			continue;
+		}
+	}
+
+	if (state != DULL) {
+		argv = realloc(argv, sizeof(char *) * (*argc + 1));
+		argv[(*argc)++] = start_of_word;
+	}
+
+	argv = realloc(argv, sizeof(char *) * (*argc + 2));
+	argv[(*argc)] = NULL;
+
+	return argv;
+}
+
+static char *shell_path(void)
+{
+	char * shells[] = { "/bin/sh", "/system/bin/sh", "/bin/bash", "/usr/local/bin/bash" };
+	for (int i = 0; i < COUNT_OF(shells); i++) {
+		if (access(shells[i], X_OK) == 0) {
+			return shells[i];
+		}
+	}
+	return NULL;
+}
+
+static void exec_child(struct procmgr *mgr,
+    const char *file, struct process_options *opts)
+{
+	const char *sh = shell_path();
+	char *args = NULL;
+
+	ev_loop_fork(EV_DEFAULT);
+	ev_loop_destroy(EV_DEFAULT_UC);
+
+	const char *process_name = file;
+	if (opts) {
+		if (opts->cwd != NULL && chdir(opts->cwd)) {
+			abort();
+		}
+		if (opts->user != NULL) {
+			switch_user(opts->user);
+		}
+		if (opts->env != NULL) {
+			environ = opts->env;
+		}
+		if (opts->process_name) {
+			process_name = opts->process_name;
+		}
+	}
+
+	if (opts && opts->args) {
+		if (asprintf(&args, "%s %s", process_name, opts->args) <= 0) {
+			abort();
+		}
+	} else {
+		args = strdup(process_name);
+	}
+
+	char **env = NULL;
+	size_t nenv = 0;
+	char *def_env[] = {
+		"PATH=/usr/local/sbin:"
+		       "/usr/local/bin:"
+		       "/usr/sbin:"
+		       "/usr/bin:"
+		       "/sbin:"
+		       "/bin:"
+		       "/usr/games:"
+		       "/usr/local/games:"
+		       "/system/bin:"
+		       "/system/sbin:"
+		       "/system/xbin:",
+		"USERNAME=nobody",
+		"HOME=/",
+		"LANG=en_US",
+		NULL
+	};
+
+	if (getenv("PATH") == NULL) {
+		env = def_env;
```

Comment:
Sounds like something we'd be debugging in a couple of years because it was too clever for some system. The man page advises against it too:

"Direct access can be made through the global variable environ, though it is recommended that changes to the enviroment still be made through the environment routines."